### ISSUE TITLE
[KRaft] Scaling of controller nodes

### ISSUE THREAD

**Issue by scholzj** [strimzi/strimzi-kafka-operator#9429]
Scaling of KRaft controller-only nodes currently doesn't seem to work. Below are described the situations and issues:

## Scale-up

The scale-up seems to currently work like this:

1. The controller node pool is scaled up
2. First, the new controller nodes are started with the new voter configuration
3. Next, the old controller nodes are rolled with the new voter configuration
4. As the old controller nodes are rolled, the leader-controller is shifting and it seems it typically ends up on one of the new nodes.
   At that point, all broker nodes seem to fail with the following error as they are not yet rolled and do not know the new leader controller.
   ```
    2023-12-04 14:16:50,669 ERROR Encountered fatal fault: Unexpected error in raft IO thread (org.apache.kafka.server.fault.ProcessTerminatingFaultHandler) [kafka-2001-raft-io-thread]
    java.lang.IllegalStateException: Cannot transition to Follower with leaderId=4 and epoch=48 since it is not one of the voters [0, 1, 2]
	    at org.apache.kafka.raft.QuorumState.transitionToFollower(QuorumState.java:382)
	    at org.apache.kafka.raft.KafkaRaftClient.transitionToFollower(KafkaRaftClient.java:522)
	    at org.apache.kafka.raft.KafkaRaftClient.maybeTransition(KafkaRaftClient.java:1575)
	    at org.apache.kafka.raft.KafkaRaftClient.maybeHandleCommonResponse(KafkaRaftClient.java:1532)
	    at org.apache.kafka.raft.KafkaRaftClient.handleFetchResponse(KafkaRaftClient.java:1113)
	    at org.apache.kafka.raft.KafkaRaftClient.handleResponse(KafkaRaftClient.java:1609)
	    at org.apache.kafka.raft.KafkaRaftClient.handleInboundMessage(KafkaRaftClient.java:1735)
	    at org.apache.kafka.raft.KafkaRaftClient.poll(KafkaRaftClient.java:2310)
	    at kafka.raft.KafkaRaftManager$RaftIoThread.doWork(RaftManager.scala:64)
	    at org.apache.kafka.server.util.ShutdownableThread.run(ShutdownableThread.java:130)
   ```
5. After rolling all the controllers, the operator rolls the brokers to introduce the new voter configuration and gets them working properly again

This seemed to work for me this was several times for scaling from 3->5 and 3->4->5.

I guess we could improve this by adding the new nodes after rolling the brokers? E.g.:
1. Do the rolling => roll the controllers to expect more voters and roll the brokers to expect more voters
2. Add the new nodes

If that is done in steps without breaking the quorum, it will likely work without causing the error in the broker nodes?

The scaling up of mixed nodes seems similar. But because the brokers are controllers as well, they do not error as with dedicated controller nodes.

## Scale-down

Scale-down currently works like this (in theory only):
1. Remove the controllers to be scaled down
2. Roll the remaining controllers to reconfigure the voters
3. Roll the brokers to reconfigure the voters

Scale-down from 5->4 nodes seems to work fine by following the steps above as it does not break the quorum. However, scale-down from 4->3 nodes will break the quorum:
1. You remove the 4th controllers
2. You need to roll the remaining 3 controllers
3. But they are configured with 4 voters so quorum needs 3 nodes online and as we have only 3 nodes left, we cannot shutdown any of them to roll it / update the voters to 3.

As a result, it gets stuck with the following error:
```
2023-12-04 17:13:06 WARN  KafkaQuorumCheck:98 - Reconciliation #1609(timer) Kafka(myproject/my-cluster): No valid lastCaughtUpTimestamp is found for controller 3
```

Scale-down of mixed nodes seems to go similarly. It looks like there is a race condition and sometimes the _deleted_ node is still seen as in-sync and the first node is rolled. But it gets stuck on the next in that case there the _controller 3_ has again no valid timestamp.

Would removing the controllers only at the end help? Can the `KafkaRoller` force this despite breaking the quorum when there is a scale-down?

## Next steps

Implementing these things could be quite complicated (even assuming the delayed creation / removal of the scaled nodes really works as I did not test it). Is it required for the GA of KRaft in Strimzi? Should we wait for [KIP-853](https://cwiki.apache.org/confluence/display/KAFKA/KIP-853%3A+KRaft+Controller+Membership+Changes) to be implemented that might allow us to change the controller quorums dynamically?

**Comment by showuon** [comment]
I agree we don't need to workaround it for now because [KIP-853](https://cwiki.apache.org/confluence/display/KAFKA/KIP-853%3A+KRaft+Controller+Membership+Changes) is still under discussion in Kafka community.

**Comment by ppatierno** [comment]
I agree, we should wait for a proper fix in Kafka upstream. Making it working with workaround and/or manual intervention is more than an hack, losing all the automation that the operator should provide in a such operation like scaling up/down.

> Is it required for the GA of KRaft in Strimzi?

Controllers are replacing ZooKeeper nodes. I was wondering how many users need to scale up/down a ZooKeeper ensemble today. I don't think that many. Of course, right now you can scale ZooKeeper if needed. Without KIP-853 you could not scale controllers properly, so it would be a not parity on features. If by GA we want to provide features parity we should wait otherwise I am fine to go GA without it.

**Comment by scholzj** [comment]
Discussed in the community call on 14.12.: Should be fixed, but it seems like it would make sense to wait for the KIP-853.

**Comment by MikeNikolayev** [comment]
KIP-853 was resolved. kafka 3.9 is released.

**Comment by scholzj** [comment]
No, KIP-853 was not really done in Kafka 3.9.0. While dynamic quorum configuration is supported now, there is no migration between the static and dynamic quorum. So our choice is either to keep waiting for Kafka 4.x to implement it and use static quorum in the meantime ... or start using the dynamic quorum and abandon all the already existing Kafka clusters. We decided to wait as we did not want to abandon our users just because of Kafka's mess.

**Comment by siegenthalerroger** [comment]
For reference (correct me if there's other jira issues more relevant) this Kafka fix is required: https://issues.apache.org/jira/browse/KAFKA-17241

in particular this subtask: https://issues.apache.org/jira/browse/KAFKA-16538

**Comment by conilas** [comment]
it seems that KAFKA-16538 has been merged and will come in kafka 4.1.0. I know this is very fresh and all ([apparently merged last week](https://github.com/apache/kafka/pull/19416)), but I wanted to check one thing:

currently, if I have 2 kafka node pools, one that has only controller role and another with a broker role: is there a way to merge them? 

one option would be to create a new kafkanodepool with controller + broker node, rebalance stuff with CruiseControl and then scale down/delete the old single-role node pools. I'm guessing this is impossible now until new versions of Strimzi adopt this (indeed very recent) change in kafka that will enable the migration from kraft 0 to 1 (and with that I guess static to dynamic).

**Comment by scholzj** [comment]
> currently, if I have 2 kafka node pools, one that has only controller role and another with a broker role: is there a way to merge them?

You can add the broker role to your controller node pool, reassign all partitions there and remove the old brokers once empty. That is the easiest way how to migrate to mixed nodes. If you have any more questions, starting a separate discussion might be bets instead of continuing under this issue.

**Comment by bollmann** [comment]
Hi, 

We're currently on Strimzi 0.45.1 and Kafka 3.9.1 and recently migrated to KRaft. Unfortunately, we misconfigured our controller kafkanodepool consisting of 3 nodes: We didn't distribute the different kafka controllers into separate availability zones and would like to fix this in hindsight. I wonder what the current best practice is in Strimzi to:

1. Remove a Kafka controller-only broker from the kafkanodepool ?
2. Add it again (this time setting the proper K8s topologySpreadConstraints so that it gets launched in the correct AZ on AWS) ?

Any insights would be highly appreciated!

**Comment by scholzj** [comment]
@bollmann It might be best to start discussion - https://github.com/orgs/strimzi/discussions - that would be a better place to discuss it without poluting the issue. Also, you should share more info on your infrastructure, storage, etc. There might be a similar solution to https://github.com/orgs/strimzi/discussions/11997 ... but that was local storage, so slightly different situation.

**Comment by ppatierno** [comment]
I assigned this issue to myself to highlight I am working on it, writing down a proposal and developing some POC. It will be based on the newly Apache Kafka versions supported by Strimzi but could be wait for Apache Kafka 4.2.0 to work properly.

**Comment by ppatierno** [comment]
Together with @mimaison, we have been heavily testing the possibility to scale controllers when using static quorum.
The findings are strictly related to how Kafka works and how to scale controllers when static quorum is used, by taking into account it was not born with this purpose (even because nodes rolling are needed and nothing is handled dynamically).
Following the results and a potential impact on Strimzi.

**Scale Up**

The new controller is added; it starts and becomes an observer by having the full list of controllers, including itself, as voters. It will join the quorum when the new leader has it in the voters list.  
The brokers should be rolled first, to get the up-to-date quorum voters list (with also the newly added controller). 
The pre-existing controllers are rolled with the updated quorum voters list that includes the newly added controller. 
The leader controller is rolled last.
This process is tested and worked when adding up to `floor((n – 1) / 2)` new controllers at a time. 

**Scale Down**

The existing controllers, except the leader and the controller to be removed, are rolled one by one to get the updated quorum voters list (without the controller that is going to be removed).  
The controller to be removed is shut down. It is done before rolling the leader, to avoid that such controller becomes leader. 
Finally, the leader is rolled to get the updated quorum voters list (without the controller that is going to be removed). 
Only at the end, the brokers are rolled with the updated quorum voters list (without the removed controller). We need to roll them after controllers, to avoid that if the leader crashes and the controller to be removed is elected as leader, they don’t know about it because we rolled them earlier as first step by updating the quorum voters list not containing the controller to remove. 

This is the Kafka golden path and safest way to execute scale down with the constraint that scaling happens one controller at time. 

**Thoughts on Strimzi** 

Rolling order: 
* When scaling up controllers, the brokers should be rolled first. It’s the opposite of what happens today within the KafkaRoller logic. 
* When scaling down controllers, the brokers should be rolled last as usual within the KafkaRoller logic. 

Limiting the controller scaling:  

* to one at the time or ...  
* ... adding up to floor ((n – 1) / 2) new controllers. 

Interleave controllers rolling and shutdown: 

* On scale down, interleaving the shutdown of the controller to be removed within the rolling controller's logic. It must happen before leader is rolled but after all the other controllers are rolled. 

We can think of solving some of the above problems by: 

* Running the scaling across two (or more?) reconciliations, for example to delay rolling brokers (for scale up) and to delay shutdown controller to remove (for scale down). Tracking status would be needed. 
* Providing to KafkaRoller the logic to skip rolling a node, in this case the leader. 
* More thoughts need how to deal with scaling down when a KafkaNodePool is deleted. Not supported? Using a finalizer?

**Comment by katheris** [comment]
@ppatierno thanks for writing up your findings. On the points about Kafka Roller, it currently orders controllers first, but since that is done upfront, wouldn't it require a simple change their to change the order? Or were you hoping to change the rolling order without making changes to KafkaRoller?

Also the KafkaRoller only rolls the pods it is told to roll, so I'm not following the need to add logic for skipping rolling a node?

More generally it's worth noting that @tinaselenge and I are likely to get back to the [KafkaRoller proposal](https://github.com/strimzi/proposals/pull/103) soon so if we need bigger KafkaRoller changes then perhaps those could be done via that.

**Comment by scholzj** [comment]
What are the implications for mixed nodes?

**Comment by ppatierno** [comment]
> @ppatierno thanks for writing up your findings. On the points about Kafka Roller, it currently orders controllers first, but since that is done upfront, wouldn't it require a simple change their to change the order? Or were you hoping to change the rolling order without making changes to KafkaRoller?

@katheris The idea, but just an idea not proven with a POC, is to avoid minimize changes in the roller to change the rolling order but doing the rolling across multiple reconciliations as mentioned in my description. So it would need first reconciliation would not roll controllers by skipping them but brokers. They will be rolled on the next reconciliation. I haven't put that much thought tough because the goal was just to describe the flow within Kafka, then the process would need a proposal to discuss if we decide to go this way and not the dynamic quorum way.

> Also the KafkaRoller only rolls the pods it is told to roll, so I'm not following the need to add logic for skipping rolling a node?

It will be told to skip the controller leader when needed (by the procedure above)

> More generally it's worth noting that @tinaselenge and I are likely to get back to the https://github.com/strimzi/proposals/pull/103 soon so if we need bigger KafkaRoller changes then perhaps those could be done via that.

It depends on time frame. How long KafkaRoller 2.0 will take to land, when we want to have the scaling controllers features, and so on.

> What are the implications for mixed nodes?

@scholzj Running the tests with combined nodes, I have got the same outcome. The procedure above seems to work the same way.

**Comment by scholzj** [comment]
> [@scholzj](https://github.com/scholzj) Running the tests with combined nodes, I have got the same outcome. The procedure above seems to work the same way.

Right. But what about changing mixed nodes to dedicated brokers/controllers or the other way around.

**Comment by ppatierno** [comment]
> Right. But what about changing mixed nodes to dedicated brokers/controllers or the other way around.

If I got it right but maybe you could clarify, one way to look at this is about:

* Using [Transitioning to separate broker and controller roles](https://strimzi.io/docs/operators/latest/deploying#proc-splitting-node-pools-roles-str) where you move from combined to dedicated nodes and then you can use the procedure we tested to scale down/up the single controller role pool.
* Or before running [Transitioning to dual-role nodes](https://strimzi.io/docs/operators/latest/deploying#proc-joining-node-pools-roles-str) you can scale down/up the single controller role pool (with the procedure we tested) and then doing the transition.

(I am referring to Strimzi doc and node pools but the same would apply to bare Kafka)

Or maybe you are referring different scenarios where:

* Start from combined nodes, then adds the same number of dedicated controllers (so it means you are actually scaling up) but then you remove the controller role from the combined nodes (back to the initial number of controllers by scaling down).
* Start from dedicated nodes. Add the controller role to brokers (actually scaling up) but then shut down the dedicated controller nodes.

But I am just guessing to interpret your comment as much as possible. Could you elaborate a little bit more what are the use cases you have in mind, please? Thanks! :-)

**Comment by ppatierno** [comment]
> Or before running [Transitioning to dual-role nodes](https://strimzi.io/docs/operators/latest/deploying#proc-joining-node-pools-roles-str) you can scale down/up the single controller role pool (with the procedure we tested) and then doing the transition.

Actually this would not make much sense before you would end up with more or less brokers you want, depending how much you scale down/up the dedicated controller nodes before the transitioning.

So yes, please clarify the use cases you have in mind.

**Comment by scholzj** [comment]
I think my question was touching pretty much all of the points. Today, we can independently on Kafka:
* Add broker role to controllers
* Remove broker role from controllers

But in your write up (https://github.com/strimzi/strimzi-kafka-operator/issues/9429#issuecomment-5164830138), you only touched scale-up and scale-down. Not:
* Adding controller role to broker nodes
* Removing controller role from broker nodes
* Adding new mixed nodes
* Removing mixed nodes

So I wonder if you had some thoughts on that.

**Comment by ppatierno** [comment]
I finally went through tests taking into account mixed nodes.

> Adding new mixed nodes
Removing mixed nodes

I already tested them in the past and reported in one of my previous comments. Following the Kafka procedure I depicted, these scenarios work. Making it working in Strimzi is still related to adhere to such procedure as described here https://github.com/strimzi/strimzi-kafka-operator/issues/9429#issuecomment-5164830138

> Adding controller role to broker nodes

It seems to work fine with Kafka on bare metal when you add the controller role on a single broker at time and follow the procedure which will be like a scale up from a Kafka perspective. Of course, it's not the same with Strimzi: when you add a "controller" role to a KafkaNodePool, you are not just adding the role to one broker but all brokers of that KafkaNodePool and they will be rolled one by one in one phase. It's not like bare metal Kafka where you can add controller role to just one broker node. It kind of "works" eventually but as said, for example, broker-0 rolled first with new "controller" role with all quorum voters (so also broker-1 and broker-2 which are not controllers yet, because not rolled yet) temporary logs the following which makes sense to me:

```shell
2026-09-14 07:18:24 ERROR [kafka-0-raft-io-thread] KafkaRaftClient:2728 - [RaftManager id=0] Unexpected error VOTER_NOT_FOUND in UPDATE_RAFT_VOTER response: InboundResponse(correlationId=31, data=UpdateRaftVoterResponseData(throttleTimeMs=0, errorCode=127, currentLeader=CurrentLeader(leaderId=5, leaderEpoch=1, host='my-cluster-controller-5.my-cluster-kafka-brokers.myproject.svc', port=9090)), source=my-cluster-controller-5.my-cluster-kafka-brokers.myproject.svc:9090 (id: 5 rack: null isFenced: false))
```

> Removing mixed nodes

As before, it seems to work fine with Kafka on bare metal when you remove the controller role from a single broker and follow the procedure which will be like a scale down from a Kafka perspective.  Of course, again, it's not the same with Strimzi: when you remove a "controller" role from a KafkaNodePool, you are removing such role from several combined nodes. Looking at the scaling down steps you are going to roll all nodes one by one while the removed controller should happen before rolling brokers (and leaders). Here Strimzi doesn't work at all. All dedicated controllers are rolled first, and we break the quorum. When it comes to roll the first combined node, it keeps just logging this error without becoming ready:

```shell
[BrokerLifecycleManager id=0] Unable to register the broker because the RPC got timed out before it could be sent.
```

So, said that, for me we are going in circle and into the wrong direction to invest (waste) time for something which was not built to work in this scenarios. This is exactly the goal of the dynamic quorum which allows to dynamically add/remove controllers without rolling nodes. For me, we should leave static quoum and focus on the dynamic quorum proposal where the POC was working fine and we should try to address the disaster recovery scenario (i.e. DNS name changes, ...) that was mentioned there.
